// ngb-toast/esm5/lib/ngb-toast.module.js
import { __decorate } from "tslib";
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { NgbToastComponent } from './component/ngb-toast-component';
import { NgbToastContainer } from './pages/ngb-toast-container';
import { NgbToastService } from './provider/ngb-toast.service';
var NgbToastModule = /** @class */ (function () {
    function NgbToastModule() {
    }
    NgbToastModule_1 = NgbToastModule;
    NgbToastModule.forRoot = function () {
        return {
            ngModule: NgbToastModule_1,
            providers: [NgbToastService]
        };
    };
    var NgbToastModule_1;
    NgbToastModule = NgbToastModule_1 = __decorate([
        NgModule({
            declarations: [
                NgbToastContainer,
                NgbToastComponent
            ],
            imports: [
                BrowserModule
            ],
            providers: [NgbToastService],
            exports: [
                NgbToastContainer,
            ],
        })
    ], NgbToastModule);
    return NgbToastModule;
}());
export { NgbToastModule };
