// ngb-toast/esm5/lib/pages/ngb-toast-container.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { NgbToastService } from '../provider/ngb-toast.service';
var NgbToastContainer = /** @class */ (function () {
    function NgbToastContainer(ngbToastService) {
        this.ngbToastService = ngbToastService;
        this.toasts = [];
    }
    NgbToastContainer.prototype.ngOnInit = function () {
        var _this = this;
        this.ngbToastService.getAllToasts().subscribe(function (toasts) {
            _this.toasts = toasts.slice().reverse();
        });
    };
    NgbToastContainer.ctorParameters = function () { return [
        { type: NgbToastService }
    ]; };
    NgbToastContainer = __decorate([
        Component({
            selector: 'ngb-toast-container',
            template: "<ngb-toast-component *ngFor=\"let toast of toasts\" [toast]=\"toast\"></ngb-toast-component>",
            styles: [":host{position:fixed;top:0;right:0;margin:.5em;z-index:1000}"]
        })
    ], NgbToastContainer);
    return NgbToastContainer;
}());
export { NgbToastContainer };
