// ngx-bootstrap-datetime-popup/esm2020/lib/ngx-bootstrap-datetime-popup.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { BsDatepickerConfig, BsDatepickerModule, DatepickerModule } from 'ngx-bootstrap/datepicker';
import { BsDropdownConfig, BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TimepickerModule } from 'ngx-bootstrap/timepicker';
import { DatetimePopupComponent } from './ngx-bootstrap-datetime-popup.component';
import { OffClickDirective } from './offclick.directive';
import * as i0 from "@angular/core";
export class DatetimePopupModule {
}
DatetimePopupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.0", ngImport: i0, type: DatetimePopupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DatetimePopupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.0", ngImport: i0, type: DatetimePopupModule, declarations: [DatetimePopupComponent, OffClickDirective], imports: [CommonModule,
        FormsModule,
        DatepickerModule,
        TimepickerModule,
        BsDatepickerModule,
        BsDropdownModule], exports: [DatetimePopupComponent] });
DatetimePopupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.0", ngImport: i0, type: DatetimePopupModule, providers: [BsDatepickerConfig, BsDropdownConfig], imports: [[
            CommonModule,
            FormsModule,
            DatepickerModule,
            TimepickerModule,
            BsDatepickerModule,
            BsDropdownModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.0", ngImport: i0, type: DatetimePopupModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DatetimePopupComponent, OffClickDirective],
                    imports: [
                        CommonModule,
                        FormsModule,
                        DatepickerModule,
                        TimepickerModule,
                        BsDatepickerModule,
                        BsDropdownModule,
                    ],
                    exports: [DatetimePopupComponent],
                    providers: [BsDatepickerConfig, BsDropdownConfig]
                }]
        }] });
