// ngx-bootstrap-datetime-popup/esm2020/lib/offclick.directive.mjs
import { Directive, EventEmitter, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class OffClickDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.offClick = new EventEmitter();
    }
    ngOnInit() {
        this.eventHandler = ($event) => {
            $event.stopPropagation();
            if (!this.elementRef.nativeElement.contains($event.target) && $event.target.localName !== 'button') {
                this.offClick.emit(null);
            }
        };
        document.addEventListener('mouseup', this.eventHandler);
    }
    ngOnDestroy() {
        document.removeEventListener('mouseup', this.eventHandler);
    }
}
OffClickDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.0", ngImport: i0, type: OffClickDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
OffClickDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.0", type: OffClickDirective, selector: "[offClick]", outputs: { offClick: "offClick" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.0", ngImport: i0, type: OffClickDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[offClick]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { offClick: [{
                type: Output
            }] } });
