// ngx-otp-input/esm2020/lib/ngx-otp-input.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class NgxOtpInputService {
    init2DArray(size) {
        return new Array(size).fill(new Array());
    }
    toArray(value) {
        return Array.isArray(value) ? value : [value];
    }
    addItemToAll(source, items) {
        if (source?.length > 0) {
            return source.map((entry) => entry.concat(items));
        }
    }
    removeItemFromAll(source, items) {
        if (source?.length > 0) {
            return source.map((entry) => entry.filter((item) => !items.includes(item)));
        }
    }
    addItemAtIndex(source, index, items) {
        if (source?.length > 0) {
            source[index] = source[index].concat(items);
            return source;
        }
    }
    removeItemAtIndex(source, index, items) {
        if (source?.length > 0) {
            source[index] = source[index].filter((item) => !items.includes(item));
            return source;
        }
    }
}
NgxOtpInputService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.1", ngImport: i0, type: NgxOtpInputService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
NgxOtpInputService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.1", ngImport: i0, type: NgxOtpInputService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.1", ngImport: i0, type: NgxOtpInputService, decorators: [{
            type: Injectable
        }] });
