// ngx-otp-input/esm2020/lib/ngx-otp-input.module.mjs
import { NgModule } from '@angular/core';
import { NgxOtpInputComponent } from './component/ngx-otp-input.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { PatternDirective } from './pattern.directive';
import { NgxOtpInputService } from './ngx-otp-input.service';
import * as i0 from "@angular/core";
export class NgxOtpInputModule {
}
NgxOtpInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.1", ngImport: i0, type: NgxOtpInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxOtpInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.1", ngImport: i0, type: NgxOtpInputModule, declarations: [NgxOtpInputComponent, PatternDirective], imports: [ReactiveFormsModule, FormsModule, CommonModule], exports: [NgxOtpInputComponent] });
NgxOtpInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.1", ngImport: i0, type: NgxOtpInputModule, providers: [NgxOtpInputService], imports: [ReactiveFormsModule, FormsModule, CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.1", ngImport: i0, type: NgxOtpInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxOtpInputComponent, PatternDirective],
                    imports: [ReactiveFormsModule, FormsModule, CommonModule],
                    exports: [NgxOtpInputComponent],
                    providers: [NgxOtpInputService],
                }]
        }] });
