// ngx-otp-input/esm2020/lib/pattern.directive.mjs
import { Directive, HostListener, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class PatternDirective {
    constructor() {
        // allowed keys apart from numeric characters
        this.allowedKeys = [
            'Backspace',
            'ArrowLeft',
            'ArrowRight',
            'Escape',
            'Tab',
        ];
    }
    onKeyDown(e) {
        if (this.allowedKeys.includes(e.key) ||
            (e.code === 'KeyA' && e.ctrlKey === true) || // Allow: Ctrl+A
            (e.code === 'KeyC' && e.ctrlKey === true) || // Allow: Ctrl+C
            (e.code === 'KeyV' && e.ctrlKey === true) || // Allow: Ctrl+V
            (e.code === 'KeyX' && e.ctrlKey === true) || // Allow: Ctrl+X
            (e.code === 'KeyA' && e.metaKey === true) || // Cmd+A (Mac)
            (e.code === 'KeyC' && e.metaKey === true) || // Cmd+C (Mac)
            (e.code === 'KeyV' && e.metaKey === true) || // Cmd+V (Mac)
            (e.code === 'KeyX' && e.metaKey === true) // Cmd+X (Mac)
        ) {
            return; // let it happen, don't do anything
        }
        else if (!this.pattern.test(e.key)) {
            e.preventDefault();
        }
    }
}
PatternDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.1", ngImport: i0, type: PatternDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
PatternDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.1", type: PatternDirective, selector: "[ngxOtpPattern]", inputs: { pattern: ["ngxOtpPattern", "pattern"] }, host: { listeners: { "keydown": "onKeyDown($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.1", ngImport: i0, type: PatternDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxOtpPattern]',
                }]
        }], propDecorators: { pattern: [{
                type: Input,
                args: ['ngxOtpPattern']
            }], onKeyDown: [{
                type: HostListener,
                args: ['keydown', ['$event']]
            }] } });
